File: src/main/java/edu/stanford/w5/ArrayHeap.java
package edu.stanford.w5;

import org.apache.commons.lang3.tuple.Pair;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.function.Function;

import static edu.stanford.common.CommonUtils.list;
import static edu.stanford.common.CommonUtils.pair;

public class ArrayHeap<K extends Comparable<K>, V> implements Heap<K, V> {

    private final LinkedList<Pair<K, V>> nodes = new LinkedList<>();

    @Override
    public void add(K key, V value) {
        int last = nodes.size();
        nodes.addLast(pair(key, value));
        bubbleUp(last);
    }

    private void bubbleUp(int childIdx) {
        Pair<K, V> child = nodes.get(childIdx);
        int parentIdx = parent(childIdx);
        Pair<K, V> parent = nodes.get(parentIdx);

        if (child.getKey().compareTo(parent.getKey()) < 0) {
            swap(childIdx, parentIdx);
            bubbleUp(parentIdx);
        }
    }

    @Override
    public Pair<K, V> extractMin() {
        Pair<K, V> node = nodes.pollFirst();
        if (node != null) {
            nodes.addFirst(nodes.pollLast());
            bubbleDown(0);
            return node;
        } else {
            return null;
        }

    }

    @Override
    public V extractMinValue() {
        return extractMin().getValue();
    }

    private void bubbleDown(int i) {
        for (Integer child : orderedChildren(i)) {
            if (isFirstKeyLess(child, i)) {
                swap(child, i);
                bubbleDown(child);
                return;
            }
        }
    }

    private List<Integer> orderedChildren(int i) {
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left > nodes.size() - 1) {
            return list();
        } else if (right > nodes.size() - 1) {
            return list(left);
        }

        return isFirstKeyLess(left, right) ? list(left, right) : list(right, left);
    }

    @Override
    public void addAll(Collection<Pair<K, V>> pairs) {
        pairs.forEach(pair -> add(pair.getKey(), pair.getValue()));
    }

    @Override
    public void addAll(Function<V, K> keySupplier, Collection<V> values) {
        values.forEach(v -> add(keySupplier.apply(v), v));
    }

    @Override
    public V delete(K key) {
        int delIndex = -1;
        for (int i = 0; i < nodes.size(); i++) {
            if (nodes.get(i).getKey().equals(key)) {
                delIndex = i;
                break;
            }
        }

        if (delIndex >= 0) {
            Pair<K, V> node = nodes.get(delIndex);
            swap(delIndex, nodes.size() - 1);
            bubbleDown(delIndex);
            return node.getValue();
        } else {
            return null;
        }
    }

    private int parent(int i) {
        return (int) Math.floor(i / 2);
    }

    private void swap(int left, int right) {
        Collections.swap(nodes, left, right);
    }

    private boolean isFirstKeyLess(int first, int second) {
        return nodes.get(first).getKey().compareTo(nodes.get(second).getKey()) < 0;
    }

    @Override
    public String toString() {
        return "ArrayHeap{" +
                "nodes=" + nodes +
                '}';
    }
}

File: src/main/scala/edu/stanford/w5/ArrayHeap.scala
package edu.stanford.w5

import scala.annotation.tailrec

trait HeapOrder[K] {
  def order(child: K, parent: K)(implicit ordering: Ordering[K]): Boolean
}

trait MinHeapOrder[K] extends HeapOrder[K] {
  def order(childKey: K, parentKey: K)(implicit ordering: Ordering[K]): Boolean = {
    ordering.lt(childKey, parentKey)
  }
}

trait MaxHeapOrder[K] extends HeapOrder[K] {
  def order(childKey: K, parentKey: K)(implicit ordering: Ordering[K]): Boolean = {
    ordering.gt(childKey, parentKey)
  }
}

class MinArrayHeap[K, V]()(implicit ordering: Ordering[K]) extends ArrayHeap[K, V]() with MinHeapOrder[K]

class MaxArrayHeap[K, V](implicit ordering: Ordering[K]) extends ArrayHeap[K, V]() with MaxHeapOrder[K]

abstract class ArrayHeap[K, V]()(implicit ordering: Ordering[K]) extends Heap[K, V] {
  self: HeapOrder[K] =>

  private var length = 2

  private var count = 0

  private var nodes = new Array[(K, V)](length)

  def add(entry: (K, V)) {
    nodes(count) = entry
    bubbleUp(count)
    onNodeAdded()
  }

  private def onNodeAdded(): Unit = {
    count += 1

    if (count >= length) {
      length *= 2
      val newNodes = new Array[(K, V)](length)
      nodes.copyToArray(newNodes)
      nodes = newNodes
    }
  }

  override def peek: Option[(K, V)] = {
    if (isEmpty) {
      None
    } else {
      Some(nodes(0))
    }
  }

  override def pop: Option[(K, V)] = {
    if (!isEmpty) {
      val first = nodes(0)
      val last = nodes(count - 1)

      nodes(0) = last
      nodes(count - 1) = null
      onNodeRemoved()

      bubbleDown(0)

      Some(first)
    } else {
      None
    }
  }

  private def onNodeRemoved(): Unit = {
    count -= 1

    if (count < length / 2) {
      length /= 2
      val newNodes = new Array[(K, V)](length)
      nodes.copyToArray(newNodes)
      nodes = newNodes
    }
  }

  @tailrec
  private def bubbleUp(childIdx: Int) {
    val childKey = keyFor(childIdx)
    val parentIdx = parentOf(childIdx)
    val parentKey = keyFor(parentIdx)

    if (order(childKey, parentKey)) {
      swap(childIdx, parentIdx)
      bubbleUp(parentIdx)
    }
  }

  private def bubbleDown(i: Int) {
    orderedChildren(i)
      .find(orderKeys(_, i))
      .foreach(child => {
        swap(child, i)
        bubbleDown(child)
      })
  }

  private def orderedChildren(i: Int): List[Int] = {
    val left: Int = 2 * i + 1
    val right: Int = 2 * i + 2

    if (left > count - 1) {
      List.empty
    } else if (right > count - 1) {
      List(left)
    } else {
      List(left, right).sortWith(orderKeys)
    }
  }

  override def isEmpty: Boolean = count == 0

  override def size: Int = count

  //  todo: implement
  def delete(key: K): Option[V] = {
    throw new NotImplementedError()
    //    var delIndex: Int = -1
    //    var i: Int = 0
    //    while (i < nodes.length) {
    //      {
    //        if (nodes(i)._1 == key) {
    //          delIndex = i
    //          break 
    //        }
    //      }
    //      {
    //        i += 1
    //        i - 1
    //      }
    //    }
    //
    //    if (delIndex >= 0) {
    //      val node: (K, V) = nodes(delIndex)
    //      swap(delIndex, nodes.length - 1)
    //      bubbleDown(delIndex)
    //      Some(node._2)
    //    } else {
    //      None
    //    }
  }

  private def orderKeys(first: Int, second: Int): Boolean = {
    order(keyFor(first), keyFor(second))
  }

  private def keyFor(index: Int): K = nodes(index)._1

  private def parentOf(i: Int): Int = Math.floor(i / 2).toInt

  private def swap(left: Int, right: Int) {
    val buf = nodes(left)
    nodes(left) = nodes(right)
    nodes(right) = buf
  }

  override def toString: String = "ArrayHeap{" + "nodes=" + nodes + '}'

}
